codalab/model/bundle_model.py

```diff
@@ -1439,6 +1440,59 @@ def find_uuid(x):
             }
             connection.execute(cl_event.insert().values(info))
 
+    # Operations on the query log
+    def add_chat_log_info(self, query_info):
+        sender_user_id = query_info.get('sender_user_id')
+        recipient_user_id = query_info.get('recipient_user_id')
+        chat = query_info.get('chat')
+        worksheet_uuid = query_info.get('worksheet_uuid')
+        bundle_uuid = query_info.get('bundle_uuid')
+        with self.engine.begin() as connection:
+            info = {
+                'time': datetime.datetime.fromtimestamp(time.time()),
+                'date': datetime.datetime.fromtimestamp(time.time()).strftime('%Y-%m-%d'),
+                'sender_user_id': sender_user_id,
+                'recipient_user_id': recipient_user_id,
+                'chat': chat,
+                'worksheet_uuid': worksheet_uuid,
+                'bundle_uuid': bundle_uuid,
+            }
+            connection.execute(cl_chat.insert().values(info))
+        result = self.get_chat_log_info({'user_id': sender_user_id})
+        return result
+
+    def get_chat_log_info(self, query_info):
+        user_id1 = query_info.get('user_id')
+        if user_id1 == None:
+            return
+        user_id1 = int(user_id1)
+        limit = query_info.get('limit')
+        with self.engine.begin() as connection:
+            query = select([cl_chat.c.date, cl_chat.c.sender_user_id, cl_chat.c.recipient_user_id, cl_chat.c.chat])
+            if int(user_id1) == 0:
+                query = query.where(or_(cl_chat.c.sender_user_id == user_id1, cl_chat.c.recipient_user_id == user_id1, cl_chat.c.sender_user_id == -1, cl_chat.c.recipient_user_id == -1))
+            else:
+                query = query.where(or_(cl_chat.c.sender_user_id == user_id1, cl_chat.c.recipient_user_id == user_id1))
+            if limit != None:
+                query = query.limit(limit)
+            # query = query.order_by(cl_chat.c.id.desc())
+            rows = connection.execute(query).fetchall()
+
+            result = {}
+            for row in rows:
+                if user_id1 == 0:
+                    user_id2 = row.sender_user_id if int(row.sender_user_id) != user_id1 and int(row.sender_user_id) != -1 else row.recipient_user_id
+                else:
+                    user_id2 = row.sender_user_id if int(row.sender_user_id) != user_id1 else row.recipient_user_id
+                    # if user is Admin, group it with System so that the user see System and Admin in the same window
+                    if user_id2 == '0':
+                        user_id2 = '-1'
+                if user_id2 in result:
+                    result[user_id2].append({'chat': row.chat, 'date': row.date, 'sender_user_id': row.sender_user_id, 'recipient_user_id': row.recipient_user_id})
```

Comment:
Rewrite to avoid duplicate code: do `if user_id2 not in result: result[user_id2] = []` first